Can't build gettext-sys if any parent path has a space in it

Probably because of https://www.gnu.org/software/autoconf/manual/autoconf-2.60/autoconf.html#Special-Chars-in-Variables

I believe the solution would be to build gettext in /tmp or some other place that has a known good path, but I'm new to this project.

Here's the output of running `$ cargo build` in gettext-sys from /tmp/space man/

$ cargo build
    Updating registry `https://github.com/rust-lang/crates.io-index`
   Compiling cc v1.0.18
   Compiling gettext-sys v0.19.8 (file:///private/tmp/space%20man/gettext-rs/gettext-sys)
error: failed to run custom build command for `gettext-sys v0.19.8 (file:///private/tmp/space%20man/gettext-rs/gettext-sys)`
process didn't exit successfully: `/private/tmp/space man/gettext-rs/target/debug/build/gettext-sys-80f993ee9d914c84/build-script-build` (exit code: 101)
--- stdout
cargo:rerun-if-env-changed=X86_64_APPLE_DARWIN_GETTEXT_SYSTEM
cargo:rerun-if-env-changed=GETTEXT_SYSTEM
cargo:rustc-link-lib=framework=CoreFoundation
cargo:rustc-link-lib=dylib=iconv
cargo:rerun-if-env-changed=X86_64_APPLE_DARWIN_GETTEXT_DIR
cargo:rerun-if-env-changed=GETTEXT_DIR
cargo:rerun-if-env-changed=X86_64_APPLE_DARWIN_GETTEXT_BIN_DIR
cargo:rerun-if-env-changed=GETTEXT_BIN_DIR
cargo:rerun-if-env-changed=X86_64_APPLE_DARWIN_GETTEXT_LIB_DIR
cargo:rerun-if-env-changed=GETTEXT_LIB_DIR
cargo:rerun-if-env-changed=X86_64_APPLE_DARWIN_GETTEXT_INCLUDE_DIR
cargo:rerun-if-env-changed=GETTEXT_INCLUDE_DIR
OPT_LEVEL = Some("0")
TARGET = Some("x86_64-apple-darwin")
HOST = Some("x86_64-apple-darwin")
TARGET = Some("x86_64-apple-darwin")
TARGET = Some("x86_64-apple-darwin")
HOST = Some("x86_64-apple-darwin")
CC_x86_64-apple-darwin = None
CC_x86_64_apple_darwin = None
HOST_CC = None
CC = None
HOST = Some("x86_64-apple-darwin")
TARGET = Some("x86_64-apple-darwin")
HOST = Some("x86_64-apple-darwin")
CFLAGS_x86_64-apple-darwin = None
CFLAGS_x86_64_apple_darwin = None
HOST_CFLAGS = None
CFLAGS = None
DEBUG = Some("true")
running: "tar" "xJf" "/private/tmp/space man/gettext-rs/gettext-sys/gettext-0.19.8.1.tar.xz" "--strip-components" "1"
running: "sh" "/private/tmp/space man/gettext-rs/target/debug/build/gettext-sys-6fc3dea112695cd0/out/gettext/configure" "--without-emacs" "--disable-java" "--disable-csharp" "--disable-c++" "--disable-shared" "--enable-static" "--enable-fast-install" "--with-included-gettext" "--with-included-glib" "--with-included-libcroco" "--with-included-libunistring" "--prefix=/private/tmp/space man/gettext-rs/target/debug/build/gettext-sys-6fc3dea112695cd0/out"
checking for a BSD-compatible install... /usr/local/bin/ginstall -c
checking whether build environment is sane...
--- stderr
configure: error: unsafe srcdir value: '/private/tmp/space man/gettext-rs/target/debug/build/gettext-sys-6fc3dea112695cd0/out/gettext'
thread 'main' panicked at '
command did not execute successfully, got: exit code: 1

build script failed, must exit now', gettext-sys/build.rs:251:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.

--- a/gettext-sys/CHANGELOG.md
+++ b/gettext-sys/CHANGELOG.md
@@ -7,9 +7,13 @@
 - Checks for build tools required by GNU gettext (Dean Leggo)
 - Bindings for `wbindtextdomain` (only available on Windows) (Alexander
     Batischev)
+- Build-time dependency on `tempfile` (Alexander Batischev)
 
 ### Changed
 - Bump bundled GNU gettext to 0.21 (Alexander Batischev)
+
+### Fixed
+- Build failure when a path contains spaces (Alexander Batischev)
 
 
 

--- a/gettext-sys/Cargo.toml
+++ b/gettext-sys/Cargo.toml
@@ -16,3 +16,4 @@ gettext-system = []
 
 [build-dependencies]
 cc = "1.0"
+tempfile = "3"

--- a/gettext-sys/README.md
+++ b/gettext-sys/README.md
@@ -56,6 +56,17 @@ ways out:
 
 - `NUM_JOBS`: sets the number of parallel build jobs.
 
+- `TMPDIR` (on Unix), `TMP`, `TEMP`, `USERPROFILE` (on Windows): set the
+    parent directory for the temporary build directory.
+
+    GNU gettext uses autotools, which [don't allow some characters][chars] in
+    paths, notably a space character. To get around that, this crate performs
+    the build in a temporary directory which usually resides somewhere under
+    _/tmp_ or _C:\\Temp_. The aforementioned env vars allow you to move the
+    build directory elsewhere.
+
+    [chars]: https://www.gnu.org/software/autoconf/manual/autoconf-2.60/autoconf.html#Special-Chars-in-Variables
+
 For target-specific configuration, each of these environment variables can be
 prefixed by an upper-cased target, for example,
 `X86_64_UNKNOWN_LINUX_GNU_GETTEXT_DIR`. This can be useful in cross compilation

--- a/gettext-sys/build.rs
+++ b/gettext-sys/build.rs
@@ -1,11 +1,13 @@
 extern crate cc;
+extern crate tempfile;
 
 use std::env;
 use std::ffi::OsString;
 use std::fs;
+use std::io::ErrorKind;
 use std::path::{Path, PathBuf};
 use std::process::Command;
-use std::io::ErrorKind;
+use tempfile::tempdir;
 
 fn env(name: &str) -> Option<String> {
     let prefix = env::var("TARGET").unwrap().to_uppercase().replace("-", "_");
@@ -158,13 +160,14 @@ fn main() {
 
     let host = env::var("HOST").unwrap();
     let src = env::current_dir().unwrap();
-    let dst = PathBuf::from(env::var_os("OUT_DIR").unwrap());
+    let build_dir = tempdir().unwrap();
+    let build_dir = build_dir.path();
 
     let cfg = cc::Build::new();
     let compiler = cfg.get_compiler();
 
-    let _ = fs::create_dir(&dst.join("build"));
-    let _ = fs::create_dir(&dst.join("gettext"));
+    let _ = fs::create_dir(&build_dir.join("build"));
+    let _ = fs::create_dir(&build_dir.join("gettext"));
 
     let mut cflags = OsString::new();
     for arg in compiler.args() {
@@ -177,27 +180,25 @@ fn main() {
         cflags.push("-DLIBXML_STATIC");
     }
 
-    if !dst.join("gettext").join("configure").is_file() {
-        let mut cmd = Command::new("tar");
-        cmd.current_dir(&dst.join("gettext"))
-           .arg("xJf")
-           .arg(&src.join("gettext-0.21.tar.xz"))
-           .arg("--strip-components")
-           .arg("1");
-        if host.contains("windows") {
-            // tar confuses local path with a remote resource because of ':'
-            cmd.arg("--force-local");
-        }
-        run(&mut cmd, "tar");
+    let mut cmd = Command::new("tar");
+    cmd.current_dir(&build_dir.join("gettext"))
+       .arg("xJf")
+       .arg(&src.join("gettext-0.21.tar.xz"))
+       .arg("--strip-components")
+       .arg("1");
+    if host.contains("windows") {
+        // tar confuses local path with a remote resource because of ':'
+        cmd.arg("--force-local");
     }
+    run(&mut cmd, "tar");
 
     let mut cmd = Command::new("sh");
     cmd.env("CC", compiler.path())
        .env("CFLAGS", cflags)
        .env("LD", &which("ld").unwrap())
        .env("VERBOSE", "1")
-       .current_dir(&dst.join("build"))
-       .arg(&posix_path(&dst.join("gettext").join("configure")));
+       .current_dir(&build_dir.join("build"))
+       .arg(&posix_path(&build_dir.join("gettext").join("configure")));
 
     cmd.arg("--without-emacs");
     cmd.arg("--disable-java");
@@ -217,7 +218,7 @@ fn main() {
         cmd.arg("--enable-threads=windows");
     }
 
-    cmd.arg(format!("--prefix={}", &posix_path(&dst)));
+    cmd.arg(format!("--prefix={}", &posix_path(&build_dir)));
 
     if target != host &&
        (!target.contains("windows") || !host.contains("windows")) {
@@ -239,10 +240,20 @@ fn main() {
     run(&mut cmd, "sh");
     run(make()
                 .arg(&format!("-j{}", env::var("NUM_JOBS").unwrap()))
-                .current_dir(&dst.join("build")), "make");
+                .current_dir(&build_dir.join("build")), "make");
     run(make()
                 .arg("install")
-                .current_dir(&dst.join("build")), "make");
+                .current_dir(&build_dir.join("build")), "make");
+
+    let dst = PathBuf::from(env::var_os("OUT_DIR").unwrap());
+    let mut cmd = Command::new("cp");
+    cmd.current_dir(&build_dir)
+       .arg("-r")
+       .arg(&build_dir.join("bin"))
+       .arg(&build_dir.join("include"))
+       .arg(&build_dir.join("lib"))
+       .arg(&dst);
+    run(&mut cmd, "cp");
 
     println!("cargo:rustc-link-lib=static=intl");
     println!("cargo:rustc-link-search=native={}/lib", dst.display());

